fastcache.go

```diff
@@ -248,59 +304,189 @@ func (b *bucket) Init(maxBytes uint64) {
+		b.flusher.chunks[i] = flushChunk{
+			chunkId:    0,
+			chunk:      *array,
+			h:          make([]uint64, 0, itemsPerChunk),
+			idx:        make([]uint64, 0, itemsPerChunk),
+			gen:        make([]uint64, 0, itemsPerChunk),
+			size:       0,
+			cleanChunk: false,
+		}
+	}
+	b.flusher.chunkSynced.Store(b.flusher.chunks)
+	index := make([]flushChunkIndexItem, primeNumber.NextPrime(uint64(maxBatch)))
+	b.flusher.index.Store(index)
 }
 
 func (b *bucket) Reset() {
 	b.mu.Lock()
 	chunks := b.chunks
 	for i := range chunks {
 		putChunk(chunks[i])
-		chunks[i] = nil
+		chunks[i] = nil //
 	}
 	b.m = make(map[uint64]uint64)
-	b.idx = 0
+	b.idx.Store(0)
 	b.gen = 1
+	b.mu.Unlock()
 	atomic.StoreUint64(&b.getCalls, 0)
 	atomic.StoreUint64(&b.setCalls, 0)
 	atomic.StoreUint64(&b.misses, 0)
 	atomic.StoreUint64(&b.collisions, 0)
 	atomic.StoreUint64(&b.corruptions, 0)
-	b.mu.Unlock()
 }
 
-func (b *bucket) cleanLocked() {
+func (b *bucket) cleanLocked(idx uint64) {
 	bGen := b.gen & ((1 << genSizeBits) - 1)
-	bIdx := b.idx
+	bIdx := idx
 	bm := b.m
 	for k, v := range bm {
 		gen := v >> bucketSizeBits
-		idx := v & ((1 << bucketSizeBits) - 1)
+		idx = v & ((1 << bucketSizeBits) - 1)
 		if (gen+1 == bGen || gen == maxGen && bGen == 1) && idx >= bIdx || gen == bGen && idx < bIdx {
 			continue
 		}
 		delete(bm, k)
 	}
 }
 
-func (b *bucket) UpdateStats(s *Stats) {
-	s.GetCalls += atomic.LoadUint64(&b.getCalls)
-	s.SetCalls += atomic.LoadUint64(&b.setCalls)
+func (b *bucket) UpdateStats(s *Stats, details bool) {
+	getCalls := atomic.LoadUint64(&b.getCalls)
+	s.GetCalls += getCalls
+	if details {
+		s.BucketGetCalls = append(s.BucketGetCalls, getCalls)
+	}
+
+	setCalls := atomic.LoadUint64(&b.setCalls)
+	s.SetCalls += setCalls
+	if details {
+		s.BucketsSetCalls = append(s.BucketsSetCalls, setCalls)
+	}
 	s.Misses += atomic.LoadUint64(&b.misses)
 	s.Collisions += atomic.LoadUint64(&b.collisions)
 	s.Corruptions += atomic.LoadUint64(&b.corruptions)
+	s.DropsInQueue += atomic.LoadUint64(&b.dropsInQueue)
+	s.DroppedWrites += atomic.LoadUint64(&b.droppedWrites)
+	s.WriteQueueSize += atomic.LoadUint64(&b.writeBufferSize) + uint64(len(b.setBuf))
+	s.SetBatchCalls += atomic.LoadUint64(&b.batchSetCalls)
+	s.DuplicatedCount += atomic.LoadUint64(&b.duplicatedCount)
 
 	b.mu.RLock()
 	s.EntriesCount += uint64(len(b.m))
 	bytesSize := uint64(0)
 	for _, chunk := range b.chunks {
 		bytesSize += uint64(cap(chunk))
 	}
-	s.BytesSize += bytesSize
 	s.MaxBytesSize += uint64(len(b.chunks)) * chunkSize
 	b.mu.RUnlock()
+	s.BytesSize += bytesSize
 }
 
-func (b *bucket) Set(k, v []byte, h uint64) {
+func (b *bucket) set(kv []byte, h uint64, idx uint64, chunkIdx uint64, needClean bool) (newNeedClean bool, newIdx uint64, newChunkId uint64) {
+	chunks := b.chunks
+	idxNew := idx + uint64(len(kv))
+	chunkIdxNew := idxNew / chunkSize
+	if chunkIdxNew > chunkIdx {
+		if chunkIdxNew >= uint64(len(chunks)) {
+			idx = 0
+			idxNew = uint64(len(kv))
+			chunkIdx = 0
+			b.gen++
```

Comment:
IMO this needs some separate function with a clear name - it's hard to figure out what's going on with this `gen` thing. Or this code is forked? Shows up like a new...